accepted/00000-saltclass-external-pillar.md

```diff
@@ -0,0 +1,147 @@
+The unknown is if the PR will be ever finished or if we can invest and try to finalize it ourselves. It is possible that the work stopped only temporarily due to changes in salt releases, or that the author being external contributor chose to abandon the effort.
+
+## Performance impact
+
+Including another pillar engine can have a negative performance impact. (#TODO measure)
+
+## More complexity means more break points
+
+Generally when running more complex code, there is bigger room for bugs. SaltClass it maintained by the Salt upstream and as of now have 19 issues opened against saltclass (counting both saltclass tops and pillar modules)
+
+## Modifying core functionality
+
+Pillar system is core functionality everything else depends on. Bugs within SaltClass can affect whole product. However proposed granular introduction of SaltClass instead of ```suma_minion``` should limit impact to only specific areas, hopefully providing enough time to catch and debug issues.
+
+## Upstream breakages, abandonment
+
+There is always some risk that upstream may decide to abandon SaltClass. So far maintaining SaltClass should be cheap, due to upstream involvement, but if the upstream decides for any reason to abandon SaltClass the cost of maintenance can increase dramatically (up to impossible if upstream decides to abandon it for some deep design reasons). As a mitigation there is a possibility to use [reclass](https://reclass.pantsfullofunix.net/) upon which SaltClass is based and shares storage details (see below).
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Implement custom jinja processing of pillars
+
+The major driver for `saltclass` is the ability to preparse pillars with jinja and also to reference pillars between each other. This can be implemented within existing `suma_minion.py` external pillar as well. Using `saltclass` saves us the development time in exchange for development time on integration, but given `saltclass` is an upstream project it in theory can help with maintenance.
+
+## Use original reclass instead
+
+Instead of salt native implementation there is an option to use the original [reclass](https://reclass.pantsfullofunix.net/) with the [reclass](https://docs.saltstack.com/en/latest/ref/pillar/all/salt.pillar.reclass_adapter.html) adapter. This can be used again as just external pillar or external tops as well.
+Drawback is need of another python package and that `reclass` itself seems to be abandoned (last commit is from 2015). There is newer fork in salt-formula git repository - https://github.com/salt-formulas/reclass (last commit from January, 2019), however from the issue list this one seems to have a problem with Python 3.6 and newer.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* performance impact
+* status of the upstream refactoring
```

Comment:
I see two areas where this RfC may cause problems elsewhere:

1) Git integration. Many of our customers rely on Git for their pillar and state files when they work on their own Salt content. We need to look into this! One pointer from the web: https://thatremindsme.woodwose.net/computing/2018/08/30/Using-saltclass-from-a-gitrepository.html

2) YAML vs. JSON: Stratos seems to now be able to display JSON files with JSON schemas from Kubernetes and Cloud Foundry in a conceptually very similar way that Formulas with Forms do. I see this potentially as a V2 (or at least alternative format) for the current forms.yml approach. Also, should we decide to store the SUSE Manager pillars in PostgreSQL at some point in the future, JSON would make things easier as well as there is a native JSON type for PostgreSQL.